src/core/include/units/bits/prime.h

```diff
@@ -0,0 +1,143 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2018 Mateusz Pusz
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <optional>
+
+namespace units::detail
+{
+
+constexpr bool is_prime_by_trial_division(std::size_t n) {
+  for (std::size_t f = 2; f * f <= n; f += 1 + (f % 2)) {
+    if (n % f == 0) { return false; }
+  }
+  return true;
+}
+
+constexpr std::size_t num_primes_between(std::size_t start, std::size_t end) {
+  std::size_t n = 0;
+
+  for (auto k = start; k < end; ++k) {
+    if (is_prime_by_trial_division(k)) { ++n; }
+  }
+
+  return n;
+}
+
+template<std::size_t Start, std::size_t End>
+constexpr auto primes_between() {
+  std::array<std::size_t, num_primes_between(Start, End)> results;
+  auto iter = std::begin(results);
+
+  for (auto k = Start; k < End; ++k) {
+    if (is_prime_by_trial_division(k)) {
+      *iter = k;
+      ++iter;
+    }
+  }
+
+  assert(iter == std::end(results));
+  return results;
+}
```

Comment:
This looks like a good use case for https://github.com/lefticus/tools/blob/main/include/lefticus/tools/static_views.hpp#L81, presented at https://www.youtube.com/watch?v=ABg4_EV5L3w.
Rather than doing the work twice (once for the size, once more for the actual population), you can define a `primes_between` that returns an [`oversized_array<std::size_t>`](https://github.com/lefticus/tools/blob/main/include/lefticus/tools/static_views.hpp#L42) (until we can use `std::vector`), and `to_span([] { return primes_between(...); })` will take care of shrinking its size to fit efficiently.

Reply:
Looks appealing!  But doesn't that rely on `consteval`?  I tried using `consteval` in another PR, and I found that we're stuck supporting a compiler which doesn't support it.

Fortunately, I think the impact of doing the work twice is minimal, since the parts we do twice are generally going to be pretty small for practical purposes.